feat: adding everest-helper enhancements into the docs (#30)
citrineos/citrineos.github.io: docs/core-concepts/testing.md

Reviewer: Minor: maybe just name to "Testing with Everest?

Author: adjusted

Reviewer: May want to adjust the filename as well

Author: adjusted

---
layout: doc
title: Testing (Everest)
---

# Testing

In the case you don't have a charger that supports OCPP 2.0.1 to experiment with, we can recommend using the Linux 
Foundation Energy project EVerest. [See here](https://github.com/EVerest) for the repository.  They have built an open source version of
charger firmware and also allow for using it as a simulator. They support OCPP 2.0.1 which makes it a great testing 
opportunity with CitrineOS. For the long route of setting up EVerst you can follow their documentation and build 

> the project yourself.[See here for Docs](https://everest.github.io/latest/general/03_quick_start_guide.html)

Reviewer: ```suggestion
the project yourself. [See here for Docs](https://everest.github.io/latest/general/03_quick_start_guide.html)
```

Author: adjusted

# Running Everest
In order to alleviate some of the complexities that may arise when starting Everest, we have created
some helpful commands that should help in getting the Everest charger simulator running locally and targeting
CitrineOS.

> You will notice in `/Server/everest` directory the new files created to support running Everest within Docker.

Reviewer: I would avoid this wording where you say "new files" as these files may be new now, but will not be so new as time goes on.

I would just word it as though the files have always been there.

Author: adjusted

In addition, we created some helpful NPM commands:

- `npm run start-everest`
- and
- `npm run start-everest-windows`

Both of which in essence do the same thing which is to trigger the `docker compose up` command from within

> the `/Server/everest` directory so that it can pick up the `Dockerfile` and the `docker-compose.yml` files:

Reviewer: Not sure the colon at the end of this sentence makes sense.

I think you may be missing a one line preamble for the code snippet below.

```shell

> EVEREST_IMAGE_TAG=0.0.16 EVEREST_TARGET_URL=ws://host.docker.internal:8081/cp001 docker-compose up

Reviewer: Honestly, you can probably just get rid of this code snippet, and then maybe adjust the wording below to read as:

"In these NPM start commands, you will notice that there are two args..."

Author: hmm, my initial thought was to include this in here explicitly showing them that they can also just run this command directly without having to use npm, maybe it is helpful to see?

Reviewer: But this code snippet isnt a command you can run?

Reviewer: Oh wait, didn't see the docker compose up at the end

```

You will notice that there are two args that are configurable:

> - `EVEREST_IMAGE_TAG` - determines the image tag that will be used for the Everest image (ghcr.io/everest/everest-demo/manager)
> - `EVEREST_TARGET_URL` - the URL that Everest will point to. Defaulting to `host.docker.internal` assuming CitrineOS will run on
>   same machine, since `localhost` won't work within Docker

Reviewer: ```suggestion
- `EVEREST_IMAGE_TAG` - The image tag that will be used for the Everest image (ghcr.io/everest/everest-demo/manager).
- `EVEREST_TARGET_URL` - The URL that Everest will point to. Defaulting to `host.docker.internal` assuming CitrineOS will run on
  same machine, since `localhost` won't work within Docker.
```

Author: adjusted

After running `npm run start-everest` (or the Windows alternative), you should see 3 running Everest containers
and the `manager` container should have the appropriate Everest logs.

### Everest UI
Now that the 3 containers are running in Docker, you should be able to navigate to `[localhost|ip]:1880/ui/` to view
the Everest simulator UI. There, you should be able to simulate the pause/resume and plug/unplug events among others.

### Everest NodeRed

> You can also view the Everest NodeRed UI `[localhost|ip]:1880/`, but it is not advisable to make any adjustments here

Reviewer: Your call, but I would almost prefer to just have this URL only contain the localhost so that I can copy and paste directly. I imagine most people will be running on local machine.

unless you have a good understanding of this configuration.

### Viewing OCPP logs in Everest
To view the OCPP logs in Everest, we have utilized Node `http-server`, which you will see being initialized
in the Dockerfile. We initialize a simple HTTP server on port `8888` and expose this port so that it is
mapped in the compose file allowing you to navigate to `localhost:8888`. This HTTP server is configured to
serve the contents of the `/tmp/everest_ocpp_logs` which is where Everest stores the OCPP logs in the
Docker container. Conveniently, the logs are in HTML format, so we can easily view them in the browser.

# Running Everest Manually
You can also use their demo repository that hosts a Docker packaged EVerest image. [See here for Github Repo](https://github.com/EVerest/everest-demo)

> To get EVerest running on the side while developing and making changes, you can follow the steps below.

Reviewer: Minor: inconsistent with usage of "EVerest" vs "Everest"

Author: adjusted 👍 thank you 🙇

> 1. Run your citrine instance locally with `docker compose up -d` in the Citrine repository

Author: adjusted

> 1. Clone the [EVerest Demo](https://github.com/EVerest/everest-demo) repository and `cd` into the repo

Author: adjusted

> 1. With citrine running excute an "add charger" script at `./citrineos/add-charger.sh` This adds a charger, location and password for the charge to Citrine

Reviewer: ```suggestion
1. With citrine running execute an "add charger" script at `./citrineos/add-charger.sh` This adds a charger, location, and password for the charger to Citrine
```
